### PR TITLE
[IMP] mail: format_float

### PR DESCRIPTION
**Pull request by fmdl**
**Description of the issue/feature this PR addresses:**
In email template to format a float there are no built-in function.

`format_float` can help you, it format with the good thousand separator and the good decimal separator (if not it use the print of python). You can give the `precision_rouding`. You can also use `uom` to add the unit name with an 'no-break-space'. 

`format_float` is also added in email template.

@nim-odoo
@Yenthe666 
@sla-subteno-it 

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by fmdl** [comment]
OPW #2299524

**Comment by fmdl** [comment]
ping @simongoffin

**Comment by fmdl** [comment]
what do you think ? @mart-e

**Comment by mart-e** [comment]
> what do you think ? @mart-e

A localised version of `float_repr`? I am not against the idea (obviously in master version) but cc @tde-banana-odoo for the mail part.

File: odoo/tools/misc.py [hunk 1]
```diff
@@ -1363,6 +1364,28 @@ def format_duration(value):
...
 
+def format_float(env, value, precision_rounding=None, uom=None, lang_code=False):
+    if precision_rounding is not None:
+        rounding = precision_rounding
+    elif uom is not None:
+        rounding = uom.rounding
+    else:
+        rounding = 0.01
+
+    dp = int(math.ceil(math.log10(1 / rounding))) if 0 < rounding < 1 else 0
```
Comment (tde-banana-odoo) [inline odoo/tools/misc.py:1375]: Whut ?

File: odoo/tools/misc.py [hunk 2]
```diff
@@ -1363,6 +1364,28 @@ def format_duration(value):
     return '%02d:%02d' % (hours, minutes)
 
 
+def format_float(env, value, precision_rounding=None, uom=None, lang_code=False):
```
Comment (tde-banana-odoo) [inline odoo/tools/misc.py:1367]: Could be great to have a docstring and some tests, especially if you want it to land in stable

File: odoo/tools/misc.py [hunk 3]
```diff
@@ -1363,6 +1364,28 @@ def format_duration(value):
...
+    if precision_rounding is not None:
+        rounding = precision_rounding
+    elif uom is not None:
```
Comment (tde-banana-odoo) [inline odoo/tools/misc.py:1370]: This is a bit strange as tools should be as model independent as possible. I would simply allow a rounding to format_float, and caller has to either give it manually or either from an uom.

I know some other tools use records (like currency), but currency is defined in base while uom in uom.

File: odoo/tools/misc.py [hunk 4]
```diff
@@ -1363,6 +1364,28 @@ def format_duration(value):
...
+    dp = int(math.ceil(math.log10(1 / rounding))) if 0 < rounding < 1 else 0
+
+    fmt = u'%.{0}f'.format(dp)
+    lang = get_lang(env, lang_code)
+
+    formated_value = lang.format(fmt,
+                                 float_round(value, precision_rounding=rounding),
+                                 grouping=True, monetary=False)
+
+    if uom is not None and uom.name:
```
Comment (tde-banana-odoo) [inline odoo/tools/misc.py:1384]: I would maybe give an uom_name parameter and use if it defined, still to be model independent. This is a simple tool that should not depend on unit of measure.

**Review by tde-banana-odoo — COMMENTED** [review]
Some comments !

File: odoo/tools/misc.py (hunk 3 above)
Comment (fmdl) [reply odoo/tools/misc.py]: good idea for `uom_name`

**Comment by fmdl** [comment]
@mart-e and @tde-banana-odoo move in master, and add test.

**Comment by mart-e** [comment]
We already have helpers to format the float for QWeb [`ir.qweb.fields.float.value_to_html`](https://github.com/odoo/odoo/blob/4a9636ce1a137f94b2a916e4b9402be5e8e5c88e/odoo/addons/base/models/ir_qweb_fields.py#L166) which does +/- the same as you (but different). For instance, it uses `precision_digits` but you use `precision_rounding`.
If we are going to introduce a similar method, let's try to match the signature and behaviour (and potentially fix what's wrong in `value_to_html`)

**Comment by fmdl** [comment]
@mart-e I have include some feature of value_to_html in format_float.

**Comment by fmdl** [comment]
@mart-e I have also imp `float_amount`

**Comment by fmdl** [comment]
ping @tde-banana-odoo

**Comment by fmdl** [comment]
ping @mart-e @tde-banana-odoo  @xmo-odoo

**Comment by fmdl** [comment]
Close by new qweb email.

**Pull request closed without merging**